### PR TITLE
Fixed #27445 -- RadioSelect widget does not work for NullBooleanField.

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket #27445](https://code.djangoproject.com/ticket/27445)

There was discussion recently on the [mailing list](https://groups.google.com/forum/m/#!searchin/django-developers/nullbooleanfield/django-developers/pSvFXcUUkRs) and it was decided that `forms.NullBooleanField` should be kept. I therefore think we can progress to fix this ticket. I've investiaged this so far but there are a few options and so I am looking for some guidance. 

The issue is that `NullBooleanField` doesn't set `choices` -  it relies on the `choices` within the `NullBooleanSelect` widget. To allow other widgets (e.g. Radio) we should set choices on the field and the widget. The PR that I've opened does this and therefore allows other widgets to be used. 

However... Should we allow choices descriptions to be customised? Ticket [#23681](https://code.djangoproject.com/ticket/23681) discusses this and there is even a suggestion to deprecate `NullBooleanSelect`. 

The tricky bit with changing names on `NullBooleanField` is `RadioSelect` / `Select` needs ('', 'True', 'False') as the first items in the choice tuples, but `NullBooleanSelect()` needs ('unknown', 'true', 'false'). Hence I can see why just using a select was sugested work, but I don't know about deprecating the current widget 😬 

In summary I think this patch solves part of the use case whilst maintaining backward compatibility. I think other options may be technically better but appear to change existing functionality.

Appreciate thoughts and guidance.

(p.s. The last example on ticket #23681 shows passing names into a `model.NullBooleanField`. Whilst this field is deprecated the same works for `model.BooleanField`. However, this gives a `TypedChoiceField`, not a `NullBooleanField` with updated names)

### CODE REVIEW

**Comment by felixxm** [comment]
> The issue is that `NullBooleanField` doesn't set `choices` - it relies on the `choices` within the `NullBooleanSelect` widget. To allow other widgets (e.g. Radio) we should set choices on the field and the widget. The PR that I've opened does this and therefore allows other widgets to be used.

We already allow `choices` customization in `models.BooleanField`. Providing a default `choices` for `BooleanField`, e.g.
```diff
diff --git a/django/db/models/fields/__init__.py b/django/db/models/fields/__init__.py
index 0fd69059ee..6e6bac28e0 100644
--- a/django/db/models/fields/__init__.py
+++ b/django/db/models/fields/__init__.py
@@ -941,6 +941,14 @@ class BooleanField(Field):
     }
     description = _("Boolean (Either True or False)")
 
+    def __init__(self, *args, **kwargs):
+        default_choices = (
+            (True, _('Yes')),
+            (False, _('No')),
+        )
+        kwargs.setdefault('choices', default_choices)
+        super().__init__(*args, **kwargs)
+
     def get_internal_type(self):
         return "BooleanField"
```
could fix also ticket-27917 :thinking:  (we have other related tickets, e.g. ticket-31636).

**Comment by smithdc1** [comment]
hmm, this is tricky. 

The patch that Mariusz provided results in every `models.BooleanField` returning a `TypedChoiceField`. I don't think this is desired. 

@jacobtylerwalls this is somewhat related to ticket #23130. Have you given any further thought to this ticket (nervous of potentially duplicating effort!). FWIW I'm not close to an answer yet.

**Comment by jacobtylerwalls** [comment]
@smithdc1 thanks for reaching out. I like the approach in your patch. I agree turning every `models.BooleanField` into a `TypedChoiceField` is not needed.

With your patch, I can use the defaults you have supplied to create the desired radio selections, and I can also use the existing functionality to rename the choices if I want:

```
class MyModel(models.Model):
    f = models.BooleanField(choices=((1, "YEP"), (2, "NOPE")), null=True)

class MyForm(forms.ModelForm):
    class Meta:
        model = MyModel
        fields = ['f']
        widgets = {'f': forms.RadioSelect}

myf = MyForm()
print(myf.as_p())
```
<p><label for="id_f_0">F:</label> <ul id="id_f">
    <li><label for="id_f_0"><input type="radio" name="f" value="" required id="id_f_0" checked>
 ---------</label>

</li>
    <li><label for="id_f_1"><input type="radio" name="f" value="1" required id="id_f_1">
 YEP</label>

</li>
    <li><label for="id_f_2"><input type="radio" name="f" value="2" required id="id_f_2">
 NOPE</label>

</li>
</ul></p>


So as I see it, there is already a way to use `RadioSelect` with `TypedChoiceField` by setting `choices` on `models.BooleanField(null=False)`, so in asking for it to be used with `NullBooleanField` this ticket is actually asking for defaults to be provided without having to manually set `choices`. I think your patch does that in a minimally invasive way.

Then, in terms of ticket-23130, I think I'm even more inclined to just not allow that combination of arguments where `choices` are provided with no `default` and `null=False`. The docs say clearly in two place (models and forms) that it takes both `choices` and `default` to suppress the blank choice. The issue arises when the blank choices can't be validated when `null=False`. We could raise a `TypeError`/system check for this combination of arguments, or rewire things so that this causes `required=True` on the form, but I'm skeptical of that route. Any thoughts for me on that?

In sum, I think these two tickets can move ahead separately.

**Comment by jacobtylerwalls** [comment]
Actually, this was super clarifying. I left some notes on ticket-23130 about how I view that now.

**Comment by smithdc1** [comment]
Hi @felixxm 

After ~~some~~ much further thought, I'm adding this back into the review queue.

As you mention there are a few tickets related to `nullboolean` type widgets:

- #23130 which Jacob has looked at is about when `null=False` and `blank=True` this leads to the situation where there is a failure at the db level. 

- This ticket is to allow the widget for `BooleanField(null=True)` to be changed when `choices` is not used.

- #27917 is very similar to this but is about being able to change the widget in the admin. I think the answer should be a change in the admin, rather than a change at the model. How this would work as yet I'm unsure, but I feel that changing the type of field for many people to have increased flexibility in the admin (less people) is the wrong solution. (I've been able to capture `BooleanField(null=True)` in `BaseModelAdmin.formfield_for_dbfield` but have yet to work out how then change the widget)

I've created a small test project to demonstrate this. With this patch when `choices` is not used and `null=True` then the widget can be changed by using something like this in the `ModelForm`. 

``` python
        widgets = {
            'both_true': RadioSelect(),
        }
```
https://github.com/smithdc1/django_test_project

File: django/forms/fields.py [hunk 1]
```diff
@@ -734,6 +734,15 @@ class NullBooleanField(BooleanField):
     """
     widget = NullBooleanSelect
 
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not isinstance(self.widget, NullBooleanSelect):
+            self.choices = self.widget.choices = [
```
Comment (carltongibson) [inline django/forms/fields.py:740]: The assignment to `self.choices` has no effect. Not being a subclass of `ChoiceField`, it's not used in validation. (Nothing fails removing it.)

**Review by carltongibson — COMMENTED** [review]
Hi @smithdc1. 

I don't think this is the right approach we're pulling knowledge about the widget up into the field... 😬


Your tests pass just with this: 

```
        f = NullBooleanField(widget=RadioSelect(choices=[
                ('', _('Unknown')),
                (True, _('Yes')),
                (False, _('No')),
            ]))
```

i.e. you configure the choices on the widget, and AFAICS it _Just Works™_. 

Do we need to do any more that just document that? (See my comment on Jacob's PR: https://github.com/django/django/pull/13437/files#r498126902)

Thanks!

**Comment by carltongibson** [comment]
Hi @smithdc1 . Thanks for your work here. I'm going to resolve with just the docs tweak as per the discussion on #13437. I think that's sufficient. 🥇
